crates/subspace-fraud-proof/src/invalid_state_transition_proof.rs

```diff
@@ -172,31 +173,130 @@ impl<'a> FetchRuntimeCode for RuntimCodeFetcher<'a> {
+    PBlock,
+    C,
+    B,
+    Exec,
+    Spawn,
+    Hash,
+    PreStateRootVerifier,
+> {
     client: Arc<C>,
     backend: Arc<B>,
     executor: Exec,
     spawn_handle: Spawn,
+    pre_state_root_verifier: PreStateRootVerifier,
     _phantom: PhantomData<(PBlock, Hash)>,
 }
 
-impl<PBlock, C, B, Exec: Clone, Spawn: Clone, Hash> Clone
-    for InvalidStateTransitionProofVerifier<PBlock, C, B, Exec, Spawn, Hash>
+impl<PBlock, C, B, Exec: Clone, Spawn: Clone, Hash, PreStateRootVerifier: Clone> Clone
+    for InvalidStateTransitionProofVerifier<PBlock, C, B, Exec, Spawn, Hash, PreStateRootVerifier>
 {
     fn clone(&self) -> Self {
         Self {
             client: self.client.clone(),
             backend: self.backend.clone(),
             executor: self.executor.clone(),
             spawn_handle: self.spawn_handle.clone(),
+            pre_state_root_verifier: self.pre_state_root_verifier.clone(),
             _phantom: self._phantom,
         }
     }
 }
 
-impl<PBlock, C, B, Exec, Spawn, Hash>
-    InvalidStateTransitionProofVerifier<PBlock, C, B, Exec, Spawn, Hash>
+/// Verifies the `pre_state_root` in [`InvalidStateTransitionProof`].
+pub trait VerifyPreStateRoot {
+    /// Returns `true` if `pre_state_root` is valid.
+    fn verify_pre_state_root(
+        &self,
+        domain_id: DomainId,
+        receipt_hash: H256,
+        pre_state_root: H256,
+        execution_phase: &ExecutionPhase,
+    ) -> bool;
+}
+
+/// Verifier of `pre_state_root` in [`InvalidStateTransitionProof`].
+///
+/// `client` could be either the primary chain client or system domain client.
+pub struct PreStateRootVerifier<Client, Block> {
+    client: Arc<Client>,
+    _phantom: PhantomData<Block>,
+}
+
+impl<Client, Block> Clone for PreStateRootVerifier<Client, Block> {
+    fn clone(&self) -> Self {
+        Self {
+            client: self.client.clone(),
+            _phantom: self._phantom,
+        }
+    }
+}
+
+impl<Client, Block> PreStateRootVerifier<Client, Block> {
+    /// Constructs a new instance of [`PreStateRootVerifier`].
+    pub fn new(client: Arc<Client>) -> Self {
+        Self {
+            client,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<Client, Block> VerifyPreStateRoot for PreStateRootVerifier<Client, Block>
+where
+    Block: BlockT,
+    Client: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+    Client::Api: ReceiptsApi<Block, domain_runtime_primitives::Hash>,
+{
+    fn verify_pre_state_root(
+        &self,
+        domain_id: DomainId,
+        receipt_hash: H256,
+        pre_state_root: H256,
+        _execution_phase: &ExecutionPhase,
+    ) -> bool {
+        let trace = self
+            .client
+            .runtime_api()
+            .execution_trace(self.client.info().best_hash, domain_id, receipt_hash)
+            .unwrap_or_default();
+
+        // TODO: Get trace index from execution phase
+        let trace_index_of_pre_state_root = 0usize;
```

Comment:
Just to be clear, for `ExecutionPhase::InitializeBlock` the `pre_state_root` is the state root of the parent block, do you mean we need to add that state root to the execution trace?

Reply:
No, the state roots are already stored in pallet-receipts, we can retrieve it from there for InitializeBlock.

Reply:
I'm not sure if we need this and the runtime api altogether. If we can already retrieve the pre_state_root since we know the parent block, wouldn;t it simpler if we just use the one from runtime and the field from fraud proof ?